### ISSUE TITLE
Bug: unexpected Overflow

### ISSUE THREAD

**Issue by eldesh** [polyml/polyml#128]
`Array.update` raises `Overflow` exception in some case, but `Subscript` should be raised for the case.

## Environment

- Poly/ML 5.8 release
- Linux x86_64

## Reproduce

```sml
$ poly
Poly/ML 5.8 Release
> val a: int array = Array.fromList [];
val a = fromList[]: int array
> Array.update (a, ~1, 0);
Exception- Subscript raised
> Array.update (Array.fromList [], ~1, 0);
Exception- Overflow unexpectedly raised while compiling

Exception- Overflow unexpectedly raised while compiling
```

**Comment by eldesh** [comment]
The same bug occurs in `Array2`.

```sml
> structure A = Array2;
structure A: ARRAY2
> val array = A.array (1, 1, 0) ;
val array = fromList[[0]]: int A.array
> A.sub (array, ~1, 0) ;
Exception- Subscript raised
> let val array = A.array (1, 1, 0) in A.sub (array, ~1, 0) end;
Exception- Overflow unexpectedly raised while compiling

Exception- Overflow unexpectedly raised while compiling
```

**Comment by dcjm** [comment]
This should have been fixed in 5369884.  The problem was actually that the Overflow was happening in the compiler itself when compiling code that would be unreachable at run-time.  Thanks for reporting it.

**Comment by eldesh** [comment]
I have confirmed that this issue has been fixed in the head of master branch 0ad5aa8 .
Thank you.
